src/main/java/com/oath/halodb/HaloDBInternal.java

```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    //TODO: use fine-grained lock if possible
+    public synchronized boolean takeSnapshot() {
+        logger.info("Start generating the snapshot");
+
+        if (this.isTombstoneFilesMerging) {
+            logger.error("DB is not ready for snapshot now");
+            return false;
+        }
+
+        try {
+            final int currentWriteFileId;
+            compactionManager.pauseCompactionThread();
+
+            // Only support one snapshot now
+            // TODO: support multiple snapshots if needed
+            File snapshotDir = getSnapshotDirectory();
+            if (snapshotDir.exists()) {
+                logger.warn("The snapshot dir is already existed. Delete the old one.");
+                FileUtils.deleteDirectory(snapshotDir);
+            }
+
+            FileUtils.createDirectoryIfNotExists(snapshotDir);
+            logger.info("Created directory for snapshot {}", snapshotDir.toString());
+
+            writeLock.lock();
+            // TODO: Need solutions to handle the file pieces after the rollover
+            // especially there are frequently rollovers
+            try {
+                forceRollOverCurrentWriteFile();
+                currentTombstoneFile = forceRollOverTombstoneFile(currentTombstoneFile);
+
+                currentWriteFileId = currentWriteFile.getFileId();
+            } catch (IOException e) {
+                logger.warn("IO exception when rollover current write files", e);
+
+                return false;
+            } finally {
+                writeLock.unlock();
+            }
+
+            File[] filesToCopy = dbDirectory.getPath().toFile()
+                .listFiles(file -> {
+                    Matcher m = Constants.STORAGE_FILE_PATTERN.matcher(file.getName());
+                    return  m.matches() && (Integer.parseInt(m.group(1)) < currentWriteFileId);
+                });
+
+            compactionManager.forceRolloverCurrentWriteFile();
+
+            logger.info("file number need to be copied: {}", filesToCopy.length);
+
+            for (File file : filesToCopy) {
+                if (file.exists()) {
+                    Path dest = Paths.get(snapshotDir.getAbsolutePath(), file.getName());
+                    logger.debug("Create links from file {} to {}", file.getName(),
+                                 dest.toFile().getAbsoluteFile());
+                    Files.createLink(dest, file.toPath());
+                } else {
+                    logger.warn("There are something wrong when create links about file {}", file.getName());
```

Comment:
If a supposed existing file doesn't exist, there must be some external interruption. Better to fail snapshot to avoid building a corrupted snapshot